File: scenarios/java/openjdk/captioning/Captioning.java
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.lang.AutoCloseable;
import java.lang.StringBuilder;
import java.math.BigInteger;
import java.net.URI;
import java.net.URISyntaxException;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.Date;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.TimeZone;
import com.microsoft.cognitiveservices.speech.*;
import com.microsoft.cognitiveservices.speech.audio.*;

public class Captioning
{
    UserConfig m_userConfig;
    AudioStreamFormat m_format;
    PullAudioInputStreamCallback m_callback;
    PullAudioInputStream m_stream;

    static private Optional<String> GetCmdOption(List<String> args, String option)
    {
        final var index = args.indexOf(option);
        if(index > -1 && index < args.size() - 1)
        {
            // We found the option (for example, "--output"), so advance from that to the value (for example, "filename").
            return Optional.of(args.get(index + 1));
        }
        else {
            return Optional.empty();
        }
    }

    static private boolean CmdOptionExists(List<String> args, String option)
    {
        return(args.contains(option));
    }

    static private AudioStreamContainerFormat GetCompressedAudioFormat(List<String> args)
    {
        final var value = GetCmdOption(args, "--format");
        if (!value.isPresent())
        {
            return AudioStreamContainerFormat.ANY;
        }
        else
        {
            switch (value.get().toLowerCase())
            {
                case "alaw" : return AudioStreamContainerFormat.ALAW;
                case "flac" : return AudioStreamContainerFormat.FLAC;
                case "mp3" : return AudioStreamContainerFormat.MP3;
                case "mulaw" : return AudioStreamContainerFormat.MULAW;
                case "ogg_opus" : return AudioStreamContainerFormat.OGG_OPUS;
                default : return AudioStreamContainerFormat.ANY;
            }
        }
    }

    static private ProfanityOption GetProfanityOption(List<String> args)
    {
        final var value = GetCmdOption(args, "--profanity");
        if (!value.isPresent())
        {
            return ProfanityOption.Masked;
        }
        else
        {
            switch (value.get().toLowerCase())
            {
                case "raw" : return ProfanityOption.Raw;
                case "removed" : return ProfanityOption.Removed;
                default : return ProfanityOption.Masked;
            }
        }
    }

    private String TimestampFromSpeechRecognitionResult(SpeechRecognitionResult result)
    {
        final var ticksPerMillisecond = BigInteger.valueOf(10000);
        final var startTime = new Date(result.getOffset().divide(ticksPerMillisecond).longValue());
        final var endTime = new Date((result.getOffset().add(result.getDuration())).divide(ticksPerMillisecond).longValue());
        
        var format = "";
        if(m_userConfig.getUseSubRipTextCaptionFormat())
        {
            // SRT format requires ',' as decimal separator rather than '.'.
            format = "HH:mm:ss,S";
        }
        else
        {
            format = "HH:mm:ss.S";
        }
        var formatter = new SimpleDateFormat(format);
        // If we don't do this, the time is adjusted for our local time zone, which we don't want.
        formatter.setTimeZone(TimeZone.getTimeZone("GMT"));
        return String.format("%s --> %s", formatter.format(startTime), formatter.format(endTime));
    }

    private String CaptionFromSpeechRecognitionResult(int sequenceNumber, SpeechRecognitionResult result)
    {
        var caption = new StringBuilder();
        if(!m_userConfig.getShowRecognizingResults() && m_userConfig.getUseSubRipTextCaptionFormat())
        {
            caption.append(String.format("%d%s", sequenceNumber, System.lineSeparator()));
        }
        caption.append(String.format("%s%s", TimestampFromSpeechRecognitionResult(result), System.lineSeparator()));
        caption.append(String.format("%s%s%s", result.getText(), System.lineSeparator(), System.lineSeparator()));
        return caption.toString();
    }

    private void WriteToConsole(String text)
    {
        if(!m_userConfig.getSuppressConsoleOutput())
        {
            System.out.print(text);
        }
    }

    private void WriteToConsoleOrFile(String text) throws IOException
    {
        WriteToConsole(text);
        if(m_userConfig.getOutputFile().isPresent())
        {
            try
            {
                var outputFile = new FileWriter(m_userConfig.getOutputFile().get(), true);
                outputFile.write(text);
                outputFile.close();
            }
            catch(IOException e)
            {
                throw e;
            }
        }
    }

    private void Initialize(UserConfig userConfig) throws IOException
    {
        m_userConfig = userConfig;
        
        if(m_userConfig.getOutputFile().isPresent())
        {
            var outputFile = new File(m_userConfig.getOutputFile().get());
            if(outputFile.exists())
            {
                outputFile.delete();
            }
        }

        if(!m_userConfig.getUseSubRipTextCaptionFormat())
        {
            WriteToConsoleOrFile(String.format("WEBVTT%s%s", System.lineSeparator(), System.lineSeparator()));
        }
    }

    static private UserConfig UserConfigFromArgs(List<String> args, String usage) throws IllegalArgumentException
    {
        var key = GetCmdOption(args, "--key");
        if(!key.isPresent())
        {
            throw new IllegalArgumentException(String.format("Missing subscription key.%s%s", System.lineSeparator(), usage));
        }
        var region = GetCmdOption(args, "--region");
        if(!region.isPresent())
        {
            throw new IllegalArgumentException(String.format("Missing region.%s%s", System.lineSeparator(), usage));
        }
        
        return new UserConfig(
            CmdOptionExists(args, "--format"),
            GetCompressedAudioFormat(args),
            GetProfanityOption(args),
            GetCmdOption(args, "--input"),
            GetCmdOption(args, "--output"),
            GetCmdOption(args, "--phrases"),
            CmdOptionExists(args, "--quiet"),
            CmdOptionExists(args, "--recognizing"),
            GetCmdOption(args, "--threshold"),
            CmdOptionExists(args, "--srt"),
            key.get(),
            region.get()
        );
    }

    private AudioConfig AudioConfigFromUserConfig()
    {
        if(m_userConfig.getInputFile().isPresent())
        {
            if (!m_userConfig.getUseCompressedAudio())
            {
                var reader = new WavFileReader(m_userConfig.getInputFile().get());
                m_format = reader.getFormat();
                reader.Close();
            }
            else
            {
                m_format = AudioStreamFormat.getCompressedFormat(m_userConfig.getCompressedAudioFormat());
            }
            m_callback = new BinaryFileReader(m_userConfig.getInputFile().get());
            m_stream = AudioInputStream.createPullStream(m_callback, m_format);
            return AudioConfig.fromStreamInput(m_stream);
        }
        else
        {
            return AudioConfig.fromDefaultMicrophoneInput();
        }
    }

    private SpeechConfig SpeechConfigFromUserConfig()
    {
        final var speechConfig = SpeechConfig.fromSubscription(m_userConfig.getSubscriptionKey(), m_userConfig.getRegion());

        speechConfig.setProfanity(m_userConfig.getProfanityOption());

        if(m_userConfig.getStablePartialResultThreshold().isPresent())
        {
            speechConfig.setProperty(PropertyId.SpeechServiceResponse_StablePartialResultThreshold, m_userConfig.getStablePartialResultThreshold().get());
        }
        
        speechConfig.setProperty(PropertyId.SpeechServiceResponse_PostProcessingOption, "TrueText");
        
        return speechConfig;
    }

    private SpeechRecognizer SpeechRecognizerFromUserConfig()
    {
        final var audioConfig = AudioConfigFromUserConfig();
        final var speechConfig = SpeechConfigFromUserConfig();
        final var speechRecognizer = new SpeechRecognizer(speechConfig, audioConfig);
        
        if(m_userConfig.getPhraseList().isPresent())
        {
            var grammar = PhraseListGrammar.fromRecognizer(speechRecognizer);
            grammar.addPhrase(m_userConfig.getPhraseList().get());
        }
        
        return speechRecognizer;
    }

    private void RecognizeContinuous(SpeechRecognizer speechRecognizer) throws ExecutionException, InterruptedException
    {
        // This lets us modify local variables from inside a lambda.
        final var done = new boolean[] { false };
        final var sequenceNumber = new int[] { 0 };

        if(m_userConfig.getShowRecognizingResults())
        {
            speechRecognizer.recognizing.addEventListener((s, e) -> {
                if(ResultReason.RecognizingSpeech == e.getResult().getReason() && e.getResult().getText().length() > 0)
                {
                    // We don't show sequence numbers for partial results.
                    WriteToConsole(CaptionFromSpeechRecognitionResult(0, e.getResult()));
                }
                else if(ResultReason.NoMatch == e.getResult().getReason())
                {
                    WriteToConsole(String.format("NOMATCH: Speech could not be recognized.%s", System.lineSeparator()));
                }
            });
        }

        speechRecognizer.recognized.addEventListener((s, e) -> {
            if(ResultReason.RecognizedSpeech == e.getResult().getReason() && e.getResult().getText().length() > 0)
            {
                sequenceNumber[0]++;
                try
                {
                    WriteToConsoleOrFile(CaptionFromSpeechRecognitionResult(sequenceNumber[0], e.getResult()));
                }
                catch(IOException ex)
                {
                    // Error output should not be suppressed, even if suppress output flag is set.
                    System.out.println(ex.getMessage());
                    done[0] = true;
                }
            }
            else if(ResultReason.NoMatch == e.getResult().getReason())
            {
                WriteToConsole(String.format("NOMATCH: Speech could not be recognized.%s", System.lineSeparator()));
            }
        });
        
        speechRecognizer.canceled.addEventListener((s, e) -> {
            if(CancellationReason.EndOfStream == e.getReason())
            {
                WriteToConsole(String.format("End of stream reached.%s", System.lineSeparator()));
                done[0] = true; // Notify to stop recognition.
            }
            else if(CancellationReason.CancelledByUser == e.getReason())
            {
                WriteToConsole(String.format("User canceled request.%s", System.lineSeparator()));
                done[0] = true; // Notify to stop recognition.
            }
            else if(CancellationReason.Error == e.getReason())
            {
                // Error output should not be suppressed, even if suppress output flag is set.
                System.out.println(String.format("Encountered error.%sError code: %d%sError details: %s%s", System.lineSeparator(), e.getErrorCode(), e.getErrorDetails()));
                done[0] = true; // Notify to stop recognition.
            }
            else
            {
                System.out.println(String.format("Request was cancelled for an unrecognized reason: %d.%s", e.getReason(), System.lineSeparator()));
                done[0] = true; // Notify to stop recognition.
            }
        });
        
        speechRecognizer.sessionStopped.addEventListener((s, e) -> {
            WriteToConsole(String.format("Session stopped.%s", System.lineSeparator()));
            done[0] = true; // Notify to stop recognition.
        });
        
        speechRecognizer.startContinuousRecognitionAsync().get();

        while(!done[0])
        {
            Thread.sleep(500);
        }

        speechRecognizer.stopContinuousRecognitionAsync().get();        
        
        return;
    }

    public static void main(String[] args) {
        // Text block content must begin on a new line.
        final var usage = """
Usage: java -cp .;target\\dependency\\* Captioning [...]

  CONNECTION
    --key KEY                     Your Azure Speech service subscription key.
    --region REGION               Your Azure Speech service region.
                                  Examples: westus, eastus

  INPUT
    --input FILE                  Input audio from file (default input is the microphone.)
    --url URL                     Input audio from URL (default input is the microphone.)
    --format FORMAT               Use compressed audio format.
                                  Valid only with --file or --url.
                                  If this is not specified, uncompressed format (wav) is assumed.
                                  Valid values: alaw, any, flac, mp3, mulaw, ogg_opus
                                  Default value: any

  RECOGNITION
    --recognizing                 Output Recognizing results (default output is Recognized results only.)
                                  These are always written to the console, never to an output file.
                                  --quiet overrides this.

  ACCURACY
    --phrases PHRASE1;PHRASE2     Example: Constoso;Jessie;Rehaan

  OUTPUT
    --help                        Show this help and stop.
    --output FILE                 Output captions to file.
    --srt                         Output captions in SubRip Text format (default format is WebVTT.)
    --quiet                       Suppress console output, except errors.
    --profanity OPTION            Valid values: raw, remove, mask
    --threshold NUMBER            Set stable partial result threshold.
                                  Default value: 3
""";

        try
        {
            List<String> argsList = Arrays.stream(args).collect(Collectors.toList());
            if(argsList.contains("--help"))
            {
                System.out.println(usage);
            }
            else
            {
                final var captioning = new Captioning();
                final var userConfig = UserConfigFromArgs(argsList, usage);
                captioning.Initialize(userConfig);
                final var speechRecognizer = captioning.SpeechRecognizerFromUserConfig();
                captioning.RecognizeContinuous(speechRecognizer);
            }
        }
        catch(Exception e)
        {
            System.out.println(e.getMessage());
        }        
    }
}

File: scenarios/cpp/windows/captioning/captioning.cpp
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//

// NOTES:
// - Install the NuGet package Microsoft.CognitiveServices.Speech.
// - The Speech SDK on Windows requires the Microsoft Visual C++ Redistributable for Visual Studio 2019 on the system. See:
// https://docs.microsoft.com/azure/cognitive-services/speech-service/speech-sdk
// - In Windows, place the following files from the Speech SDK in the same folder as your compiled .exe:
//     - Microsoft.CognitiveServices.Speech.core.dll
//     - Microsoft.CognitiveServices.Speech.extension.audio.sys.dll

#include <exception>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <optional>
#include <sstream>
#include <speechapi_cxx.h>
#include "helper.h"

using namespace Microsoft::CognitiveServices::Speech;
using namespace Microsoft::CognitiveServices::Speech::Audio;
using namespace Microsoft::CognitiveServices::Speech::Speaker;

class Captioning
{
private:

    std::shared_ptr<UserConfig> m_userConfig = NULL;
    std::shared_ptr<AudioStreamFormat> m_format = NULL;
    std::shared_ptr<BinaryFileReader> m_callback = NULL;
    std::shared_ptr<PullAudioInputStream> m_stream = NULL;

    std::string TimestampFromSpeechRecognitionResult(std::shared_ptr<SpeechRecognitionResult> result)
    {
        std::ostringstream strTimestamp;
        auto timestamp = TimestampFromTicks(result->Offset(), result->Offset() + result->Duration());    
        
        std::ostringstream startSeconds_1, endSeconds_1;
        // setw value is 2 for seconds + 1 for floating point + 3 for decimal places.
        startSeconds_1 << std::setfill('0') << std::setw(6) << std::fixed << std::setprecision(3) << timestamp.startSeconds;
        endSeconds_1 << std::setfill('0') << std::setw(6) << std::fixed << std::setprecision(3) << timestamp.endSeconds;    
        
        if (m_userConfig->useSubRipTextCaptionFormat)
        {
            // SRT format requires ',' as decimal separator rather than '.'.
            std::string startSeconds_2(startSeconds_1.str());
            std::string endSeconds_2(endSeconds_1.str());
            replace(startSeconds_2.begin(), startSeconds_2.end(), '.', ',');
            replace(endSeconds_2.begin(), endSeconds_2.end(), '.', ',');
            strTimestamp << std::setfill('0') << std::setw(2) << timestamp.startHours << ":"
                << std::setfill('0') << std::setw(2) << timestamp.startMinutes << ":" << startSeconds_2 << " --> "
                << std::setfill('0') << std::setw(2) << timestamp.endHours << ":"
                << std::setfill('0') << std::setw(2) << timestamp.endMinutes << ":" << endSeconds_2;
        }
        else
        {
            strTimestamp << std::setfill('0') << std::setw(2) << timestamp.startHours << ":"
                << std::setfill('0') << std::setw(2) << timestamp.startMinutes << ":" << startSeconds_1.str() << " --> "
                << std::setfill('0') << std::setw(2) << timestamp.endHours << ":"
                << std::setfill('0') << std::setw(2) << timestamp.endMinutes << ":" << endSeconds_1.str();
        }
        
        return strTimestamp.str ();
    }

    std::string LanguageFromSpeechRecognitionResult(std::shared_ptr<SpeechRecognitionResult> result)
    {
        if (m_userConfig->languageIDLanguages.has_value())
        {
            auto languageIDResult = AutoDetectSourceLanguageResult::FromResult(result);
            return "[" + languageIDResult->Language + "] ";
        }
        else
        {
            return "";
        }
    }

    std::string CaptionFromSpeechRecognitionResult(int sequenceNumber, std::shared_ptr<SpeechRecognitionResult> result)
    {
        std::string caption;
        if (!m_userConfig->showRecognizingResults && m_userConfig->useSubRipTextCaptionFormat)
        {
            caption += sequenceNumber + "\n";
        }
        caption += TimestampFromSpeechRecognitionResult(result) + "\n"
            + LanguageFromSpeechRecognitionResult(result)
            + result->Text + "\n\n";
        return caption;
    }

    void WriteToConsole(std::string text)
    {
        if (!m_userConfig->suppressConsoleOutput)
        {
            std::cout << text << std::flush;
        }
    }

    void WriteToConsoleOrFile(std::string text)
    {
        WriteToConsole(text);
        if (m_userConfig->outputFile.has_value())
        {
            std::ofstream outputStream;
            outputStream.open(m_userConfig->outputFile.value(), std::ios_base::app);
            outputStream << text;
            outputStream.close();
        }
    }

    std::shared_ptr<Audio::AudioConfig> AudioConfigFromUserConfig()
    {
        if (m_userConfig->inputFile.has_value())
        {
            if (EndsWith(m_userConfig->inputFile.value(), ".wav"))
            {
                auto reader = std::make_shared<WavFileReader>(m_userConfig->inputFile.value());
                m_format = AudioStreamFormat::GetWaveFormatPCM(reader->GetFormat().SamplesPerSec, reader->GetFormat().BitsPerSample, reader->GetFormat().Channels);
                reader->Close();
            }
            else
            {
                m_format = AudioStreamFormat::GetCompressedFormat(m_userConfig->compressedAudioFormat);
            }
            m_callback = std::make_shared<BinaryFileReader>(m_userConfig->inputFile.value());
            m_stream = AudioInputStream::CreatePullStream(m_format, m_callback);
            return AudioConfig::FromStreamInput(m_stream);
        }
        else
        {
            return Audio::AudioConfig::FromDefaultMicrophoneInput();
        }
    }

    std::shared_ptr<SpeechConfig> SpeechConfigFromUserConfig()
    {
        std::shared_ptr<SpeechConfig> speechConfig;
        if (m_userConfig->languageIDLanguages.has_value())
        {
            auto endpoint = V2EndpointFromRegion(m_userConfig->region);
            speechConfig = SpeechConfig::FromEndpoint(endpoint, m_userConfig->subscriptionKey);
        }
        else
        {
            speechConfig = SpeechConfig::FromSubscription(m_userConfig->subscriptionKey, m_userConfig->region);
        }

        speechConfig->SetProfanity(m_userConfig->profanityOption);

        if (m_userConfig->stablePartialResultThreshold.has_value())
        {
            // Note: To get default value:
            // std::cout << speechConfig->GetProperty(PropertyId::SpeechServiceResponse_StablePartialResultThreshold) << std::endl;
            speechConfig->SetProperty(PropertyId::SpeechServiceResponse_StablePartialResultThreshold, m_userConfig->stablePartialResultThreshold.value());
        }
        
        speechConfig->SetProperty(PropertyId::SpeechServiceResponse_PostProcessingOption, "TrueText");
        
        return speechConfig;
    }

public:
    Captioning(std::shared_ptr<UserConfig> userConfig)
        : m_userConfig(userConfig)
    {
        if (m_userConfig->outputFile.has_value())
        {
            std::filesystem::path outputFile { m_userConfig->outputFile.value() };
            // If the output file exists, truncate it.
            if (std::filesystem::exists(outputFile))
            {
                std::ofstream outputStream;
                outputStream.open(m_userConfig->outputFile.value());
                outputStream.close();            
            }
        }
        if (!m_userConfig->useSubRipTextCaptionFormat && !m_userConfig->showRecognizingResults)
        {
            WriteToConsoleOrFile("WEBVTT\n\n");
        }
    }

    std::shared_ptr<SpeechRecognizer> SpeechRecognizerFromUserConfig()
    {
        auto audioConfig = AudioConfigFromUserConfig();
        auto speechConfig = SpeechConfigFromUserConfig();
        std::shared_ptr<SpeechRecognizer> speechRecognizer;
        
        if (m_userConfig->languageIDLanguages.has_value())
        {
            auto autoDetectSourceLanguageConfig = AutoDetectSourceLanguageConfig::FromLanguages(m_userConfig->languageIDLanguages.value());
            speechRecognizer = SpeechRecognizer::FromConfig(speechConfig, autoDetectSourceLanguageConfig, audioConfig);
        }
        else
        {
            speechRecognizer = SpeechRecognizer::FromConfig(speechConfig, audioConfig);
        }

        if (m_userConfig->phraseList.has_value())
        {
            auto grammar = PhraseListGrammar::FromRecognizer(speechRecognizer);
            grammar->AddPhrase(m_userConfig->phraseList.value());
        }
        
        return speechRecognizer;
    }

    std::optional<std::string> RecognizeContinuous(std::shared_ptr<SpeechRecognizer> speechRecognizer)
    {
        std::promise<std::optional<std::string>> recognitionEnd;
        auto sequenceNumber = 0;

        if (m_userConfig->showRecognizingResults) {
            // Capture variables we will need inside the lambda.
            // See:
            // https://www.cppstories.com/2020/08/lambda-capturing.html/
            speechRecognizer->Recognizing.Connect([this](const SpeechRecognitionEventArgs& e)
                {
                    if (ResultReason::RecognizingSpeech == e.Result->Reason && e.Result->Text.length() > 0)
                    {
                        // We don't show sequence numbers for partial results.
                        WriteToConsole(CaptionFromSpeechRecognitionResult(0, e.Result));
                    }
                    else if (ResultReason::NoMatch == e.Result->Reason)
                    {
                        WriteToConsole("NOMATCH: Speech could not be recognized.\n");
                    }
                });
        }

        speechRecognizer->Recognized.Connect([this, &sequenceNumber](const SpeechRecognitionEventArgs& e)
            {
                if (ResultReason::RecognizedSpeech == e.Result->Reason && e.Result->Text.length() > 0)
                {
                    sequenceNumber++;
                    WriteToConsoleOrFile(CaptionFromSpeechRecognitionResult(sequenceNumber, e.Result));
                }
                else if (ResultReason::NoMatch == e.Result->Reason)
                {
                    WriteToConsole("NOMATCH: Speech could not be recognized.\n");
                }
            });

        speechRecognizer->Canceled.Connect([this, &recognitionEnd](const SpeechRecognitionCanceledEventArgs& e)
            {
                if (CancellationReason::EndOfStream == e.Reason)
                {
                    WriteToConsole("End of stream reached.\n");
                    recognitionEnd.set_value(std::nullopt); // Notify to stop recognition.
                }
                else if (CancellationReason::CancelledByUser == e.Reason)
                {
                    WriteToConsole("User canceled request.\n");
                    recognitionEnd.set_value(std::nullopt); // Notify to stop recognition.
                }
                else if (CancellationReason::Error == e.Reason)
                {
                    std::ostringstream error;
                    error << "Encountered error.\n"
                        << "ErrorCode: " << (int)e.ErrorCode << "\n"
                        << "ErrorDetails: " << e.ErrorDetails << std::endl;
                    recognitionEnd.set_value(std::optional<std::string>{ error.str() }); // Notify to stop recognition.
                }
                else
                {
                    std::ostringstream error;
                    error << "Request was cancelled for an unrecognized reason: " << (int)e.Reason << std::endl;
                    recognitionEnd.set_value(std::optional<std::string>{ error.str() }); // Notify to stop recognition.
                }
            });

        speechRecognizer->SessionStopped.Connect([this, &recognitionEnd](const SessionEventArgs& e)
            {
                WriteToConsole("Session stopped.\n");
                recognitionEnd.set_value(std::nullopt); // Notify to stop recognition.
            });

        // Starts continuous recognition. Uses StopContinuousRecognitionAsync() to stop recognition.
        speechRecognizer->StartContinuousRecognitionAsync().get();

        // Waits for recognition end.
        auto result = recognitionEnd.get_future().get();

        // Stops recognition.
        speechRecognizer->StopContinuousRecognitionAsync().get();
        
        return result;
    }
};

int main(int argc, char* argv[])
{
    const std::string usage = "Usage: caption.exe [...]\n\n"
"  CONNECTION\n"
"    --key KEY                     Your Azure Speech service subscription key.\n"
"    --region REGION               Your Azure Speech service region.\n"
"                                  Examples: westus, eastus\n\n"
"  LANGUAGE\n"
"    --languages LANG1,LANG2       Enable language identification for specified languages.\n"
"                                  Example: en-US,ja-JP\n\n"
"  INPUT\n"
"    --input FILE                  Input audio from file (default input is the microphone.)\n"
"    --url URL                     Input audio from URL (default input is the microphone.)\n"
"    --format FORMAT               Use compressed audio format.\n"
"                                  Valid only with --file or --url.\n"
"                                  If this is not specified, uncompressed format (wav) is assumed.\n"
"                                  Valid values: alaw, any, flac, mp3, mulaw, ogg_opus\n"
"                                  Default value: any\n\n"
"  RECOGNITION\n"
"    --recognizing                 Output Recognizing results (default output is Recognized results only.)\n"
"                                  These are always written to the console, never to an output file.\n"
"                                  --quiet overrides this.\n\n"
"  ACCURACY\n"
"    --phrases PHRASE1;PHRASE2     Example: Constoso;Jessie;Rehaan\n\n"
"  OUTPUT\n"
"    --help                        Show this help and stop.\n"
"    --output FILE                 Output captions to file.\n"
"    --srt                         Output captions in SubRip Text format (default format is WebVTT.)\n"
"    --quiet                       Suppress console output, except errors.\n"
"    --profanity OPTION            Valid values: raw, remove, mask\n"
"    --threshold NUMBER            Set stable partial result threshold.\n"
"                                  Default value: 3\n";

    try
    {
        if (CmdOptionExists(argv, argv + argc, "--help"))
        {
            std::cout << usage << std::endl;
        }
        else
        {
            auto userConfig = UserConfigFromArgs(argc, argv, usage);
            auto captioning = std::make_shared<Captioning>(userConfig);
            auto speechRecognizer = captioning->SpeechRecognizerFromUserConfig();
            auto error = captioning->RecognizeContinuous(speechRecognizer);
            if (error.has_value())
            {
                std::cout << error.value() << std::endl;
            }
        }
    }
    catch (std::exception e)
    {
        std::cout << e.what() << std::endl;
    }
}
